nixos/tests/ec2.nix

```diff
@@ -0,0 +1,77 @@
+{ system ? builtins.currentSystem }:
+
+with import ../lib/testing.nix { inherit system; };
+with import ../lib/qemu-flags.nix;
+with pkgs.lib;
+
+let
+  image =
+    (import ../lib/eval-config.nix {
+      inherit system;
+      modules = [ ../maintainers/scripts/ec2/amazon-hvm-config.nix ];
+    }).config.system.build.amazonImage;
+
+  makeEc2Test = { name, userData, script, hostname ? "ec2-instance", sshPublicKey ? null }:
+    let
+      metaData = pkgs.stdenv.mkDerivation {
+        name = "metadata";
+        buildCommand = ''
+          mkdir -p $out/2011-01-01
+          ln -s ${pkgs.writeText "userData" userData} $out/2011-01-01/user-data
+
+          mkdir -p $out/1.0/meta-data
+          echo "${hostname}" > $out/1.0/meta-data/hostname
+        '' + optionalString (sshPublicKey != null) ''
+          mkdir -p $out/1.0/meta-data/public-keys/0
+          ln -s ${pkgs.writeText "sshPublicKey" sshPublicKey} $out/1.0/meta-data/public-keys/0/openssh-key
+        '';
+      };
+    in makeTest {
+      name = "ec2-" + name;
+      nodes = {};
+      testScript =
+        ''
+          use File::Temp qw/ tempfile /;
+          my ($fh, $filename) = tempfile();
+
+          `qemu-img create -f qcow2 -o backing_file=${image}/nixos.img $filename`;
+
+          system("cd ${metaData} && ${pkgs.python}/bin/python -m SimpleHTTPServer 27182 &");
+
+          my $startCommand = "qemu-kvm -m 768 -net nic -net 'user,net=169.254.0.0/16,guestfwd=tcp:169.254.169.254:80-cmd:netcat localhost 27182'";
+          $startCommand .= " -drive file=" . Cwd::abs_path($filename) . ",if=virtio,werror=report";
+          $startCommand .= " \$QEMU_OPTS";
+         
+          my $machine = createMachine({ startCommand => $startCommand });
+          ${script}
+        '';
+    };
+
+  snakeOilPrivateKey = [
+    "-----BEGIN EC PRIVATE KEY-----"
+    "MHcCAQEEIHQf/khLvYrQ8IOika5yqtWvI0oquHlpRLTZiJy5dRJmoAoGCCqGSM49"
+    "AwEHoUQDQgAEKF0DYGbBwbj06tA3fd/+yP44cvmwmHBWXZCKbS+RQlAKvLXMWkpN"
+    "r1lwMyJZoSGgBHoUahoYjTh9/sJL7XLJtA=="
+    "-----END EC PRIVATE KEY-----"
+  ];
+
+  snakeOilPublicKey = pkgs.lib.concatStrings [
+    "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHA"
+    "yNTYAAABBBChdA2BmwcG49OrQN33f/sj+OHL5sJhwVl2Qim0vkUJQCry1zFpKTa"
+    "9ZcDMiWaEhoAR6FGoaGI04ff7CS+1yybQ= snakeoil"
+  ];
+in {
+  bootEc2NixOps = makeEc2Test {
+    name = "nixops-userdata";
+    userData = ''
+      SSH_HOST_DSA_KEY_PUB:${snakeOilPublicKey}
+      SSH_HOST_DSA_KEY:${pkgs.lib.concatStringsSep "|" snakeOilPrivateKey}
+    '';
+    hostname = "secret-hostname";
+    script = ''
+      $machine->start;
+      $machine->waitForFile("/root/user-data");
```

Comment:
I still need to test the actual desired functionality:
* that the host key is actually set
* that the ssh authorized key is set
* that the hostname is set

To do that, I'd really like to connect to the VM from the host. Is that frowned upon? Is there a better way to get the current VM test machinery to do more of my work for me?